### PR TITLE
Filtered data in filterstate UI

### LINKED ISSUE #131: [Feature Request]: Counts of the filter should be based on filtered-data
### Feature description

Histogram in RangeFilterState, or bars in the ChoicesFilterState, etc. base now on unfiltered data. Which means That if somebody adds (for example) SEX filter it always displays the same numbers of observations for M, F, U and NA. The point is to rerender these values based on the state of the filtered-data. This will help users to follow what exactly remains after applying other filters.

image

--------

Tip: this will require passing filtered-data (reactive) to the `srv_filter_panel` and pass it down to the place where `FilterState` is initialized (FilteredData$srv_filter_panel > FilteredDataset[i]$srv_add_filter_state > FilterStates[i]$srv_add_filter_state > init_filter_state). In the image below I present how srv modules are called withing filter-panel classes. `FilteredData$filter_panel_srv` calls three modules which can be seen in the filter-panel as "Summary", "Active filters", "Add filter". In the beginning active filters (`srv`) are empty and new filters can be added through `srv_add`. When new filter is selected then a new FIlterState is created and pushed to the ReactiveQueue. `FIlterStates$srv` detects that a new `FilterState` has been added and calls `insert_filter_state_ui` which calls `FilterState$ui` and `FilterState$srv` modules.

image

### LINKED ISSUE #142: [Feature request]: Counts of the levels should include filtered/unfiltered information: FilterState with slider
blocked by #131 

1. FilterState with slider should have a density which shows how much data has been filtered. The bars should show unfiltered data and what remained after filtering (5)
image
2. ~~FilterState with checkbox should show bars for filtered and unfiltered counts.~~
     - ~~F (100/231)~~
     - ~~M (50/169)~~
3. ~~FilterState with pickerInput - labels of the pickerInput should show similar counts to the (2). See the RACE filter for example~~

### LINKED ISSUE #161: [Feature request]: Counts of the levels should include filtered/unfiltered information: FilterState with checkbox
blocked by #131 

FilterState with checkbox should show bars for filtered and unfiltered counts. (5)
     - F (100/231)
     - M (50/169)

### LINKED ISSUE #162: [Feature request]: Counts of the levels should include filtered/unfiltered information: FilterState with pickerInput
blocked by #131 

FilterState with pickerInput - labels of the pickerInput should show filtered and unfiltered counts.

### PR DESCRIPTION
**Pull request by nikolas-burkoff**
Merge into filter branch:

Closes #131 and closes #142 and closes #161 and closes #162

Still to do:
- [x] Make it work with MAE
- [x] labels for logicalstate
- [x] Remove missing and infinite values from geom_histogram to stop warnings being shown
- [x] Test with more data - also does it actually work?? 
--------------------------

After merge onto feature branch:
- [ ] Make range graph align with slider (or put issue in backlog)
- [ ] Add formal tests
- [ ] Refactor?
- [ ] Docs
- [ ] Update pictures in teal, vignettes etc.
- [ ] [set labels correctly](https://github.com/insightsengineering/teal.slice/pull/164#issuecomment-1385658130) - see [comment](https://github.com/insightsengineering/teal.slice/pull/164#issuecomment-1387121156)
- [ ] [pickerinput allowing multiple selections before closing](https://github.com/insightsengineering/teal.slice/pull/164#issuecomment-1385789079) see [comment](https://github.com/insightsengineering/teal.slice/pull/164#issuecomment-1387121156)
 

Note I had to replace the smoothed density histogram with a "blocky" one as if you use the old version it ends up being very misleading as the density of the selected goes beyond your selection in the slider

image

Example app: (also test with other filter types including empty (i.e. all NA) and logical)

```
library(teal)

DF <- data.frame(
  choice = factor(rep(letters[1:10], 60), levels = letters[1:12]),
  choice2 = rep(c("M", "F", "M"), 200),
  range = c(runif(598), Inf, NA)
)

DF$choice[1:2] <- NA
DF$choice2[4:5] <- NA

app <- init(
  data = teal_data(
    dataset("DF", DF)
  ),
  modules = example_module()
)

runApp(app)

```

### CODE REVIEW

**Comment by nikolas-burkoff** [comment]
So the counts I think are not as helpful as they could be - when you deselect "M" you get 0/399:
image

But it's probably more informative to have:
image

So you can see there are 305 available given other filtering out of the 399 in total. And then checking/un-checking the checkbox does not change the label

**Comment by asbates** [comment]
Initial thoughts.

* It doesn't look like NA/Inf counts are working for numeric. If the boxes are checked, shouldn't the numbers be 1/1?
* For slider NA/Inf counts checking/unchecking one can check/uncheck the other.
* The select input closes after selecting one value. Is that supposed to happen?

image




https://user-images.githubusercontent.com/36482959/212970289-e8c3a460-d3cc-49e9-bbd8-78824ede4ca2.mp4

**Comment by nikolas-burkoff** [comment]
I've discussed with @gogonzo and if we change how we do this then we could make the labels show the more intuitive value as described [here](https://github.com/insightsengineering/teal.slice/pull/164#issuecomment-1385658130). 

```
   unfiltered data         flag matrix
row_number: x y z        x_flag  y_flag
         1  1 3 2           T       T 
         2  4 3 7           F       T
         3  1 1 1           T       T
```
filter on x where x > 1 && filter on y where y < 4

If filter y changes to y < 2 then flag matrix y_flag changes to F, F, T
this triggers an event on all (visible) filter states that the flag matrix has changed and would then trigger
an attempt to recalculate the labels (x/y) or range histogram on the visible filter states. But as only column y has changed 
this means the event to update labels/graph on filterstate y would be ignored

When creating a new filter state/ making visible the ui for an existing filter state the flag matrix
can be used to calculate which rows are filtered using only this filter state to enable the correct (x/y) label/range histogram to be shown

This would also have the advantages of no reactivity loops when quickly changing UI inputs which would solve the pickerInput issue (otherwise we could use https://github.com/dreamRs/shinyWidgets/issues/392) and also stop any reactivity loops from too quickly change UI.

This would be for another PR into the filter branch though

**Comment by gogonzo** [comment]
MAE example
```
#options(teal.log_level = "TRACE", teal.show_js_log = TRUE)
bslib::bs_theme(version = 5)
devtools::load_all("teal.slice")
devtools::load_all("teal")
library(scda)
library(SummarizedExperiment)

MAE <- hermes::multi_assay_experiment

funny_module <- function (label = "Filter states") {
  checkmate::assert_string(label)
  module(
    label = label,
    ui = function(id, ...) {
      ns <- NS(id)
      div(
        h2("The following filter calls are generated:"),
        verbatimTextOutput(ns("filter_calls")),
        actionButton(ns("reset"), "reset_to_default")
      )
    },
    server = function(input, output, session, data, filter_panel_api) {
      checkmate::assert_class(data, "tdata")
      observeEvent(input$reset, {
        set_filter_state(filter_panel_api, default_filters)
      })

      output$filter_calls <- renderText({
        filter_panel_api |> get_filter_state() |> yaml::as.yaml()
      })
    },
    filters = "all"
  )
}
mae_default_filters <- list(
  MAE = list(
    subjects = list(
      DURSUR = list(selected = c(5, 10))
    )
  )
)

mae_data <- teal.data::teal_data(teal.data::dataset("MAE", MAE))

app <- init(
  data = mae_data,
  modules = list(funny_module()),
  filter = mae_default_filters
)

runApp(app)

```

##  CDISC example
```
#options(teal.log_level = "TRACE", teal.show_js_log = TRUE)
bslib::bs_theme(version = 5)
devtools::load_all("teal.slice")
devtools::load_all("teal")
library(scda)

ADSL <- synthetic_cdisc_data("latest")$adsl
ADSL$empty <- NA
ADSL$logical1 <- FALSE
ADSL$logical <- sample(c(TRUE, FALSE), size = nrow(ADSL), replace = TRUE)
ADSL$numeric <- rnorm(nrow(ADSL))
ADSL$categorical <- sample(letters[1:10], size = nrow(ADSL), replace = TRUE)
ADSL$date <- Sys.Date() + seq_len(nrow(ADSL))
ADSL$datetime <- Sys.time() + seq_len(nrow(ADSL)) * 3600 * 12

ADSL$numeric[1:10] <- NA
ADSL$numeric[11:20] <- Inf
ADSL$logical[20:30] <- NA
ADSL$date[30:40] <- NA
ADSL$datetime[40:50] <- NA

ADTTE <- synthetic_cdisc_data("latest")$adtte
ADRS <- synthetic_cdisc_data("latest")$adrs

funny_module <- function (label = "Filter states") {
  checkmate::assert_string(label)
  module(
    label = label,
    ui = function(id, ...) {
      ns <- NS(id)
      div(
        h2("The following filter calls are generated:"),
        verbatimTextOutput(ns("filter_calls")),
        actionButton(ns("reset"), "reset_to_default")
      )
    },
    server = function(input, output, session, data, filter_panel_api) {
      checkmate::assert_class(data, "tdata")
      observeEvent(input$reset, {
        set_filter_state(filter_panel_api, default_filters)
      })

      output$filter_calls <- renderText({
        filter_panel_api |> get_filter_state() |> yaml::as.yaml()
      })
    },
    filters = "all"
  )
}

default_filters <- list(
  ADSL = list(
    logical = list(selected = FALSE),
    numeric = list(selected = c(0, 140), keep_na = TRUE, keep_inf = TRUE),
    date = list(selected = c("2023-12-18", "2050-10-12")),
    datetime = list()
  ),
  ADTTE = list(
    AVAL = list(selected = c(10, 1000), keep_na = TRUE, keep_inf = TRUE)
  )
)

data <- cdisc_data(
  cdisc_dataset("ADSL", ADSL),
  cdisc_dataset("ADTTE", ADTTE),
  cdisc_dataset("ADRS", ADRS)
)

app <- init(
  data = data,
  modules = list(funny_module()),
  filter = default_filters
)

runApp(app)
```

File: R/FilterStateRange.R [hunk 1]
```diff
@@ -39,30 +39,46 @@ RangeFilterState <- R6::R6Class( # nolint
...
 
+      pretty_range_inputs <- private$get_pretty_range_inputs(x)
+
+      private$unfiltered_histogram <-
+        ggplot2::ggplot(data.frame(x = Filter(is.finite, private$filtered_values()))) +
+        ggplot2::geom_histogram(
+          ggplot2::aes(x = x),
+          bins = 100,
+          fill = grDevices::rgb(211 / 255, 211 / 255, 211 / 255),
+          color = grDevices::rgb(211 / 255, 211 / 255, 211 / 255)
```
Comment (asbates) [inline R/FilterStateRange.R:73]: I'm not sure hard coding these colors is a good idea. If I change my theme via `bslib` for example, I want all the colors of the app to match that theme. We may be able to get the theme colors via [`bslib::bs_get_variables()`](https://rstudio.github.io/bslib/reference/bs_get_variables.html).

Although, this may be better handled as part of a broader effort to [align styling across all app components](https://github.com/insightsengineering/teal/issues/735).

**Comment by asbates** [comment]
> 

So if I understand correctly the flag matrix will be stored in one of the filter panel objects. So 

- When a filter is added, we add a new column to this matrix that is T/F depending on the value of the filter. 
- When a filter is updated, we update the column in the filter matrix.
- When a filter is removed, we remove this column in the filter matrix.
- For displaying counts of a variable, we add up the number of T/F values in the corresponding column in the filter matrix.
- Could this filter matrix also be used to make the filtered data? i.e. the data that is actually plotted, etc.

**Review by gogonzo — APPROVED** [review]
I've tested with DF and MAE in a multiple ways. I've checked that: 
- all states are properly returned
- adding and removing states works
- re setting (limited feature) works
- counts are updated in all related inputs

I think we can merge to feature branch and continue work there. @nikolas-burkoff please close the issues, we will create a new ones for extra stuff.

File: R/FilterStateRange.R (hunk 1 above)
Comment (nikolas-burkoff) [reply R/FilterStateRange.R:73]: yes I think this makes sense - there's a lot going on here so I'll add a note in the issue you mention

**Comment by pawelru** [comment]
@nikolas-burkoff this looked like a big change without NEWS update. Is that intended? If not - can you please amend that?

**Comment by pawelru** [comment]
yes you got me :) haven't looked at the target branch
